### ISSUE TITLE
sched: resource serialization failed: Resource temporarily unavailable

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-sched#336]
I believe I reproduced this error on flux-sched@master:

(after `make install` to prefix=/tmp/flux)
```
$ /tmp/flux/bin/flux start -s 4 
flux-start: warning: setting --bootstrap=selfpmi due to --size option
(flux-182416) $ for i in `seq 0 100`; do flux submit -n3 sleep 1; done
submit: Submitted jobid 1
submit: Submitted jobid 2
submit: Submitted jobid 3
submit: Submitted jobid 4
submit: Submitted jobid 5
submit: Submitted jobid 6
submit: Submitted jobid 7
2018-05-03T19:41:05.959315Z sched.err[0]: 7 resource serialization failed: Resource temporarily unavailable
2018-05-03T19:41:05.959327Z sched.err[0]: failed to request allocate for job 7
submit: Submitted jobid 8
2018-05-03T19:41:05.979984Z sched.err[0]: 8 resource serialization failed: Resource temporarily unavailable
2018-05-03T19:41:05.979997Z sched.err[0]: failed to request allocate for job 8
submit: Submitted jobid 9
2018-05-03T19:41:06.000229Z sched.err[0]: 9 resource serialization failed: Resource temporarily unavailable
2018-05-03T19:41:06.000241Z sched.err[0]: failed to request allocate for job 9
submit: Submitted jobid 10
2018-05-03T19:41:06.019617Z sched.err[0]: 10 resource serialization failed: Resource temporarily unavailable
2018-05-03T19:41:06.019629Z sched.err[0]: failed to request allocate for job 10
submit: Submitted jobid 11
2018-05-03T19:41:06.041011Z sched.err[0]: 11 resource serialization failed: Resource temporarily unavailable
2018-05-03T19:41:06.041028Z sched.err[0]: failed to request allocate for job 11
submit: Submitted jobid 12
2018-05-03T19:41:06.061536Z sched.err[0]: 12 resource serialization failed: Resource temporarily unavailable
2018-05-03T19:41:06.061548Z sched.err[0]: failed to request allocate for job 12
submit: Submitted jobid 13
2018-05-03T19:41:06.081189Z sched.err[0]: 13 resource serialization failed: Resource temporarily unavailable
2018-05-03T19:41:06.081204Z sched.err[0]: failed to request allocate for job 13
submit: Submitted jobid 14
2018-05-03T19:41:06.103026Z sched.err[0]: 14 resource serialization failed: Resource temporarily unavailable
2018-05-03T19:41:06.103040Z sched.err[0]: failed to request allocate for job 14
submit: Submitted jobid 15
2018-05-03T19:41:06.124665Z sched.err[0]: 15 resource serialization failed: Resource temporarily unavailable
2018-05-03T19:41:06.124682Z sched.err[0]: failed to request allocate for job 15
submit: Submitted jobid 16
2018-05-03T19:41:06.146217Z sched.err[0]: 16 resource serialization failed: Resource temporarily unavailable
2018-05-03T19:41:06.146231Z sched.err[0]: failed to request allocate for job 16
submit: Submitted jobid 17
2018-05-03T19:41:06.166334Z sched.err[0]: 17 resource serialization failed: Resource temporarily unavailable
2018-05-03T19:41:06.166347Z sched.err[0]: failed to request allocate for job 17
submit: Submitted jobid 18
2018-05-03T19:41:06.185942Z sched.err[0]: 18 resource serialization failed: Resource temporarily unavailable
2018-05-03T19:41:06.185956Z sched.err[0]: failed to request allocate for job 18
submit: Submitted jobid 19
2018-05-03T19:41:06.207720Z sched.err[0]: 19 resource serialization failed: Resource temporarily unavailable
2018-05-03T19:41:06.207733Z sched.err[0]: failed to request allocate for job 19
submit: Submitted jobid 20
2018-05-03T19:41:06.228120Z sched.err[0]: 20 resource serialization failed: Resource temporarily unavailable
2018-05-03T19:41:06.228142Z sched.err[0]: failed to request allocate for job 20
submit: Submitted jobid 21
2018-05-03T19:41:06.249501Z sched.err[0]: 21 resource serialization failed: Resource temporarily unavailable
2018-05-03T19:41:06.249519Z sched.err[0]: failed to request allocate for job 21
submit: Submitted jobid 22
submit: Submitted jobid 23
submit: Submitted jobid 24
submit: Submitted jobid 25
submit: Submitted jobid 26
submit: Submitted jobid 27

```

**Comment by dongahn** [comment]
@grondo: do you happened to have a broker log when this failure manifests itself?

**Comment by dongahn** [comment]
Actually, it was easy to reproduce this on quartz and I have the log file.

**Comment by dongahn** [comment]
```
2018-05-03T20:53:31.562072Z job.debug[0]: Setting job 13 to submitted
2018-05-03T20:53:31.568703Z sched.debug[0]: attempting job 13 state change from null to submitted
2018-05-03T20:53:31.568764Z sched.debug[0]: check callback about to schedule jobs.
2018-05-03T20:53:31.569459Z sched.debug[0]: Found 108 core(s) for job 13, required: 3
2018-05-03T20:53:31.569507Z sched.debug[0]: selected core0
2018-05-03T20:53:31.569519Z sched.debug[0]: selected core1
2018-05-03T20:53:31.569536Z sched.debug[0]: selected core2
2018-05-03T20:53:31.569597Z sched.err[0]: 13 resource serialization failed: Resource temporarily unavailable
2018-05-03T20:53:31.569614Z sched.err[0]: failed to request allocate for job 13
```

**Comment by dongahn** [comment]
FYI --

I believe this is due to a failure within the new lightweight serializer. `EAGAIN` (Resource temporarily unavailable) is a red herring. It seems this errno was set and recovered  but errno wasn't reset to be 0. 

Is there any good way to check where `EAGAIN` was raised?

I will keep looking at the serializer to fix the issue.

**Comment by dongahn** [comment]
Hmmm, the assumption I had in the new serializer was that the resrc_tree object selected for the job doesn't include a high-level `resrc` type that isn't a direct ancestor of the allocated `resrc` (e.g., `core`).

```
2018-05-03T21:43:47.381675Z sched.info[0]: resrc type: cluster, path: /cluster, basename: cluster, name: cluster, digest: (null), id: -1, state: invalid, uuid: 4ff0e2d2-c5d8-479f-9f5b-609c8c9723ff, size: 1, avail: 1
2018-05-03T21:43:47.381693Z sched.info[0]: resrc type: node, path: /cluster/quartz1532, basename: node, name: quartz1532, digest: 535D63A5408B1909AB75656275C2F7DC3A2149B5.0, id: 0, state: invalid, uuid: e391005d-9f64-47bd-860f-c76854b745e7, size: 1, avail: 1
2018-05-03T21:43:47.381709Z sched.info[0]: resrc type: numanode, path: /cluster/quartz1532/numanode0, basename: numanode, name: numanode0, digest: (null), id: 0, state: invalid, uuid: 46829105-e15d-45bf-b7d6-8a0c642c4d49, size: 1, avail: 1
2018-05-03T21:43:47.381724Z sched.info[0]: resrc type: numanode, path: /cluster/quartz1532/numanode1, basename: numanode, name: numanode1, digest: (null), id: 1, state: invalid, uuid: 5db6b12a-75c8-41b4-b89a-8424f6a8b002, size: 1, avail: 1
2018-05-03T21:43:47.381739Z sched.info[0]: resrc type: socket, path: /cluster/quartz1532/numanode1/socket1, basename: socket, name: socket1, digest: (null), id: 1, state: invalid, uuid: fd2654c1-7559-4821-ba14-0581b06ee964, size: 1, avail: 1
2018-05-03T21:43:47.381760Z sched.info[0]: resrc type: core, path: /cluster/quartz1532/numanode1/socket1/core27, basename: core, name: core27, digest: (null), id: 27, state: allocated, uuid: 02701c83-883e-4363-8a67-f8574da67695, size: 1, avail: 0, allocs, 98: 1
2018-05-03T21:43:47.381784Z sched.info[0]: resrc type: core, path: /cluster/quartz1532/numanode1/socket1/core28, basename: core, name: core28, digest: (null), id: 28, state: allocated, uuid: 8f3bdac5-6c90-435b-abf3-683d3fae7801, size: 1, avail: 0, allocs, 98: 1
2018-05-03T21:43:47.381800Z sched.info[0]: resrc type: core, path: /cluster/quartz1532/numanode1/socket1/core29, basename: core, name: core29, digest: (null), id: 29, state: allocated, uuid: a23e049b-f0f3-41be-b229-3af3c7965763, size: 1, avail: 0, allocs, 98: 1
```

But dumping the `resrc_tree` object for the failed job shows this isn't the case!! 

Notice `numanode0` is included when `core27`, `28` and `29` are selected.

```
2018-05-03T21:43:47.381709Z sched.info[0]: resrc type: numanode, path: /cluster/quartz1532/numanode0, basename: numanode, name: numanode0, digest: (null), id: 0, state: invalid, uuid: 46829105-e15d-45bf-b7d6-8a0c642c4d49, size: 1, avail: 1
```

**Comment by dongahn** [comment]
This may have happened because there might even be more to this previous problem: https://github.com/flux-framework/flux-core/issues/1400#issuecomment-378675739... *sigh*

**Comment by dongahn** [comment]
Looking closely [this code](https://github.com/flux-framework/flux-sched/blob/master/resrc/resrc_reqst.c#L569) once again, the parent vertex selection is done during preorder so there is no way one can prune those high-level vertices that don't have any selected resources at their subtree! That can only be determined during postorder.

To patch it, I will try to rearrange the code so that "found tree" selection is done during post order. Nevertheless, it seems clear, we need a more rigorous discipline in writing a recursive code -- work for the new `resource` code.

**Comment by dongahn** [comment]
This should be fixed in PR #333.
